keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,359 @@
+   class. These fields are out-of-scope for the initial proposal.
+
+#### API Specification
+
+_(This is a simplified declaration, syntactic details will be covered in the API PR review)_
+
+```go
+type RuntimeClass struct {
+    metav1.TypeMeta
+    // ObjectMeta minimally includes the RuntimeClass name, which is used to reference the class.
+    // Namespace should be left blank.
+    metav1.ObjectMeta
+
+    Spec RuntimeClassSpec
+}
+
+type RuntimeClassSpec struct {
+    // Parameters holds opaque parameters that are interpreted by the runtime (CRI implementation).
+    Parameters map[string]string
+}
+```
+
+The runtime is selected by the pod by specifying the RuntimeClass in the PodSpec. Once the pod is
+scheduled, the RuntimeClass cannot be changed.
+
+```go
+type PodSpec struct {
+    ...
+    // RuntimeClassName refers to a RuntimeClass object with the same name,
+    // which should be used to run this pod.
+    // +optional
+    RuntimeClassName string
+    ...
+}
+```
+
+The `legacy` RuntimeClass name is reserved. The legacy RuntimeClass is defined to be fully backwards
+compatible with current Kubernetes. This means that the legacy runtime does not specify any runtime
+parameters or perform any feature validation (all features are "supported").
+
+```go
+const (
+    // RuntimeClassNameLegacy is a reserved RuntimeClass name. The legacy
+    // RuntimeClass does not specify any runtime parameters or perform any
+    // feature validation.
+    RuntimeClassNameLegacy = "legacy"
+)
+```
+
+An unspecified RuntimeClassName `""` is equivalent to the `legacy` RuntimeClass, though the field is
+not defaulted to `legacy` (to leave room for configurable defaults in a future update).
+
+#### Runtime Parameters
+
+Runtime parameters are provided as a mechanism for passing non-portable runtime-specific arguments
+that are generally static for the RuntimeClass, but dynamically interpreted by the runtime. They
+provide a way of specifying options that are tightly coupled with the class of runtime, but should
+not be used to enumerate general pod options as this would lead to a combinatorial explosion of
+RuntimeClass definitions.
+
+Examples of runtime parameters include:
+
+```
+// Decide which underlying OCI runtime should be used. This replaces the
+// annotations used by CRI-o, containerd & frakti to select between sandboxed
+// runtimes.
+"oci-runtime": "runsc"
+
+// Change the default set of capabilities used by the runtime. This is a good
+// way of surfacing behaviors that were previously implicit in docker but not
+// well defined in Kubernetes.
+"default-capabilities": "[]"
+
+// Set the infrastructure container image. More generally, configuring options
+// not otherwise exposed in the Kubrenetes API.
+"infra-image": "example.com/self-monitoring-pod:v2"
+```
+
+_Note: RuntimeClass is intended to be consumed by a single runtime implementation, so parameter keys
+do not need to be "namespaced"._
```

Comment:
For my understanding: you could have multiple runtimeClasses on the same cluster, referring to the same runtime but with different parameters, right?  I'm thinking of, say, runc with and without SELinux, and a future when scheduling could take RuntimeClass into consideration.

Reply:
Yeah, definitely. 2 usecases I have in mind for this:

1. Same CRI runtime, different OCI runtime (not sure if you were including this)
2. "Native" RuntimeClass - This would be the default runc runtime, but applying the constraints & specs that "should" have been there from the start. The `legacy` RuntimeClass would apply to the same runc runtime, but without any constraints / supported features. (will make more sense when I follow up with the supported features addition)

To your example of runc with & without SELinux, if you're running mixed nodes where some do & don't have the SELinux LSM installed, then this makes sense. However, if you're talking about configuring a runtimeclass on a single node that supports SELinux, that feels like something that would be better configured in cluster-level policy configuration. Does that make sense?

Reply:
dumb question what is the difference between CRI runtime and OCI runtime ?

Reply:
Let me see if I understand about the cluster-level policy config: I suppose there are two ways for users to come at determining what runtime to use: 

- a "developer intent" that a given workload should really have a certain level of isolation or requires certain features from its runtime, and this is defined in the PodSpec

- a "security policy" defined cluster-wide that says that workloads, or a certain class of workloads, require a certain level of isolation or runtime features. This is defined in some cluster-wide policy - should that also refer to this RuntimeClass definition? Is the idea that this will be policed at admission time?